File: src/source/registry.zig
const std = @import("std");
const types = @import("../types.zig");
const http = @import("../util/http.zig");

pub const RegistrySource = struct {
    allocator: std.mem.Allocator,
    registry_url: []const u8,

    pub fn init(allocator: std.mem.Allocator, registry_url: []const u8) !RegistrySource {
        return .{ .allocator = allocator, .registry_url = try allocator.dupe(u8, registry_url) };
    }

    pub fn deinit(self: *RegistrySource) void {
        self.allocator.free(self.registry_url);
    }

    pub fn resolve(self: RegistrySource, a: std.mem.Allocator, name: []const u8) ![]const u8 {
        const url = try std.fmt.allocPrint(a, "{s}/{s}", .{ self.registry_url, name });
        defer a.free(url);

        var client = http.HttpClient.init(a);
        defer client.deinit();

        var resp = try client.get(url);
        defer resp.deinit(a);

        const parsed = try std.json.parseFromSlice(std.json.Value, a, resp.body, .{});
        defer parsed.deinit();

        const root = parsed.value;
        const versions = root.object.get("versions") orelse return error.PackageNotFound;
        var latest: ?[]const u8 = null;
        var latest_key: []const u8 = "";
        var iter = versions.object.iterator();
        while (iter.next()) |entry| {
            const ver = types.Version.parse(entry.key_ptr.*) catch continue;
            if (latest) |_| {
                const current = types.Version.parse(latest_key) catch continue;
                if (ver.order(current) == .gt) {
                    latest_key = entry.key_ptr.*;
                    latest = entry.key_ptr.*;
                }
            } else {
                latest_key = entry.key_ptr.*;
                latest = entry.key_ptr.*;
            }
        }

        return a.dupe(u8, latest_key);
    }

    pub fn fetch(self: RegistrySource, a: std.mem.Allocator, id: types.PackageIdentity) ![]u8 {
        var ver_buf: [64]u8 = undefined;
        const ver_str = try std.fmt.bufPrint(&ver_buf, "{}.{}.{}", .{ id.version.major, id.version.minor, id.version.patch });

        const tarball_url = try std.fmt.allocPrint(a, "{s}/{s}/-/{s}-{s}.tgz", .{ self.registry_url, id.name, id.name, ver_str });
        defer a.free(tarball_url);

        var client = http.HttpClient.init(a);
        defer client.deinit();

        var resp = try client.get(tarball_url);
        const body = resp.body;
        resp.body = &.{};
        return body;
    }
};

File: src-rs/src/source/registry.rs
use crate::source::SourceError;
use crate::types::{PackageIdentity, Version};
use crate::util::http::HttpClient;

pub struct RegistrySource {
    pub registry_url: String,
}

impl RegistrySource {
    #[must_use]
    pub fn new(registry_url: String) -> Self {
        Self { registry_url }
    }

    pub fn resolve(&self, name: &str) -> Result<String, SourceError> {
        let client = HttpClient::new();
        let url = format!("{}/{name}", self.registry_url);
        let body = client.get(&url).map_err(|e| SourceError::NetworkError(e.to_string()))?;

        let parsed: serde_json::Value =
            serde_json::from_slice(&body).map_err(|_| SourceError::PackageNotFound)?;

        let versions = parsed
            .get("versions")
            .and_then(|v| v.as_object())
            .ok_or(SourceError::PackageNotFound)?;

        let mut latest: Option<String> = None;

        for (ver_str, _) in versions {
            if let Ok(ver) = Version::parse(ver_str) {
                match &latest {
                    Some(l) => {
                        if let Ok(current) = Version::parse(l) {
                            if ver > current {
                                latest = Some(ver_str.clone());
                            }
                        }
                    }
                    None => {
                        latest = Some(ver_str.clone());
                    }
                }
            }
        }

        latest.ok_or(SourceError::VersionNotFound)
    }

    pub fn fetch(&self, identity: &PackageIdentity) -> Result<Vec<u8>, SourceError> {
        let client = HttpClient::new();
        let ver_str = format!("{}.{}.{}", identity.version.major, identity.version.minor, identity.version.patch);
        let tarball_url = format!(
            "{}/{name}/-/{name}-{ver_str}.tgz",
            self.registry_url,
            name = identity.name
        );
        let body = client.get(&tarball_url).map_err(|e| SourceError::NetworkError(e.to_string()))?;
        Ok(body)
    }
}
